From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Wed, 13 Nov 2024 17:26:19 +0100
Subject: apt: Fix matching gstreamer pkgs where the only modifier is
 ()(64bit)

When the list of GstCaps specified is empty, and the only modifier we're given
is ()(64bit), our regex eats the outer parenthesis so we're left with ")(64bit"
while trying to match "()(64bit".

This change would introduce a trailing comma when the list of specified GstCaps
is instead not empty, which gst_caps_from_string today can handle, but for
abundance of caution let's avoid it.

Test cases:
pkcon what-provides "gstreamer1(decoder-audio/mpeg)(mpegversion=4)()(64bit)"
pkcon what-provides "gstreamer1(decoder-audio/mpeg)(mpegversion=4)"
pkcon what-provides "gstreamer1(decoder-video/x-h265)()(64bit)"
pkcon what-provides "gstreamer1(decoder-video/x-h265)"

Origin: https://github.com/PackageKit/PackageKit/pull/810
Bug-Ubuntu: https://launchpad.net/bugs/2060730
---
 backends/apt/gst-matcher.cpp | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/backends/apt/gst-matcher.cpp b/backends/apt/gst-matcher.cpp
index fd5a9ca..75a342b 100644
--- a/backends/apt/gst-matcher.cpp
+++ b/backends/apt/gst-matcher.cpp
@@ -71,17 +71,21 @@ GstMatcher::GstMatcher(gchar **values)
                 if (!opt.empty()) {
                     size_t start_pos = 0;
                     // This is hardcoded in pk-gstreamer-install, so we also hardcode it here
-                    const string x86_64 = "()(64bit";
+                    const string x86_64 = ")(64bit";
 
                     if (ends_with(opt.c_str(), x86_64.c_str())) {
                             // We hardcode 64bit -> amd64 here
                             arch = "amd64";
-                            // -1 -> remove the last )
-                            opt.erase(opt.end() - x86_64.length() - 1, opt.end());
+                            opt.erase(opt.end() - x86_64.length(), opt.end());
                     }
 
                     // Replace all ")(" with "," - convert from input to serialized caps format
                     while ((start_pos = opt.find(")(", start_pos)) != string::npos) {
+                        if (start_pos == opt.length()-2) {
+                            // Avoid trailing comma
+                            opt.erase(start_pos, 2);
+                            break;
+                        }
                         opt.replace(start_pos, 2, ",");
                         start_pos++;
                     }
